### PR TITLE
Improve context detection for `PSR12.Operators.OperatorSpacing` / `Squiz.WhiteSpace.OperatorSpacing` - avoids fixer conflict, bow out on parse error

### PR DESCRIPTION
**Pull request by fredden**
## Description

I have investigated one of the issues identified in https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/152

- Standard: `PSR12`
- Test file: `src/Standards/Generic/Tests/PHP/RequireStrictTypesUnitTest.13.inc`
- Sniffs involved in the conflict:
  - `PSR12.Operators.OperatorSpacing`
  - `PSR12.Files.DeclareStatement`

> 4 | ERROR | [x] Expected at least 1 space before "="; 0 found (PSR12.Operators.OperatorSpacing.NoSpaceBefore)
> 4 | ERROR | [x] Expected no space between directive and the equals sign in a declare statement (PSR12.Files.DeclareStatement.SpaceFoundAfterDirective)

<details><summary>Verbose output from <code>phpcbf</code> run</summary>

```
	=> Fixing file: 1/1 violations remaining
---START FILE CONTENT---
1|<?php
2|
3|// Safeguard sniff handles parse error/live coding correctly.
4|declare(strict_types=
5|
--- END FILE CONTENT ---
	PSR12.Operators.OperatorSpacing:96 replaced token 6 (T_EQUAL on line 4) "=" => "[30;1m·[0m="


	=> Fixing file: 1/1 violations remaining [made 1 pass]...
	* fixed 1 violations, starting loop 2 *
---START FILE CONTENT---
1|<?php
2|
3|// Safeguard sniff handles parse error/live coding correctly.
4|declare(strict_types =
5|
--- END FILE CONTENT ---
	PSR12.Files.DeclareStatement:102 replaced token 6 (T_WHITESPACE on line 4) "[30;1m·[0m=" => "="
```

<hr></details>

In order to resolve / fix this conflict, I have taken the following steps:

- Rename existing test cases for both sniffs to allow multiple test files.
- Add new test with a parse error to each sniff (code taken from trigger test file listed above) -- see also https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/143
- Change the code in both sniffs to bow out when a parse error has been detected.

## Suggested changelog entry
Updated `PSR12.Operators.OperatorSpacing` and `PSR12.Files.DeclareStatement` to bow out when a parse error is detected.

## Related issues/external references
- https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/152

## Types of changes
- [x] Bug fix _(non-breaking change which fixes an issue)_

## PR checklist
- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.

### CODE REVIEW

**Review by jrfnl — CHANGES_REQUESTED** [review]
Hi @fredden Thanks for this PR!

I agree with your analysis of the fixer conflict and the solution direction.

I do have some remarks and questions though about the implemented fixes:
1. For both "rename test file" commits: the `switch` statement is missing the `default` case returning an empty array.
2. Again both sniffs: if no fixes are expected, there should not be a `.fixed` file.
3. For the `OperatorSpacing` sniff - which extends the Squiz `OperatorSpacing` sniff:
    - Is there a particular reason why you put the fix in the `PSR12` sniff ? (instead of in the `Squiz` sniff)
        I can imagine the same issue would also be troublesome for the Squiz sniff.
    - Is there a particular reason why you do the live coding check _after_ the `isOperator()` check ?
    - Why are you checking for live coding using `isset($tokens[($stackPtr + 2)])` instead of doing a `findNext(T_WHITESPACE, ($stackPtr + 1), null, true)` ?
        If there would be multiple new lines after the unfinished `=` your check would pass and the fixer conflict would still come into play.
4. For the `DeclareStatement` sniff:
    - Aside from the same question about the `isset()`, I see you've put the "bow out" in the section of the sniff strictly dealing with the assignment operator.
        Have you considered that the sniff should maybe not run at all if the `declare` statement is unfinished, i.e. does not have a close parenthesis ?


Just FYI: I have a WIP branch locally to do a complete rewrite of the PSR12 `DeclareStatement` sniff as there is a _lot_ more wrong with that sniff. Hoping to at some point find the time to finish the rewrite. I won't let it block this PR though.

**Comment by fredden** [comment]
> 1. For both "rename test file" commits: the `switch` statement is missing the `default` case returning an empty array.

`switch` statements do not require a `default` case. There is a `return` statement immediately following the `switch`. To appease this remark, I have moved this into the switch.

> 2. Again both sniffs: if no fixes are expected, there should not be a `.fixed` file.

These files were very useful during debugging / development. I have removed them as per your preference.

> 3. For the `OperatorSpacing` sniff - which extends the Squiz `OperatorSpacing` sniff:
>    * Is there a particular reason why you put the fix in the `PSR12` sniff ? (instead of in the `Squiz` sniff)
>      I can imagine the same issue would also be troublesome for the Squiz sniff.

I was focusing on the PSR12 standard. I have now moved the fix to the Squiz standard and added a test there to catch the same problem.

>    * Is there a particular reason why you do the live coding check _after_ the `isOperator()` check ?

I didn't want to modify anything outside of the scope I was looking at: PSR12. Given the above, the check is now _in_ the `isOperator()` check.

>    * Why are you checking for live coding using `isset($tokens[($stackPtr + 2)])` instead of doing a `findNext(T_WHITESPACE, ($stackPtr + 1), null, true)` ?
>      If there would be multiple new lines after the unfinished `=` your check would pass and the fixer conflict would still come into play.

I don't have a good justification for this. I've applied the suggested change.

> 4. For the `DeclareStatement` sniff:
>    * Aside from the same question about the `isset()`, I see you've put the "bow out" in the section of the sniff strictly dealing with the assignment operator.
>      Have you considered that the sniff should maybe not run at all if the `declare` statement is unfinished, i.e. does not have a close parenthesis ?

This sounds like a good improvement that could be made to the sniff. I don't think it's in scope here, but I've make a code change to cover this to address this feedback.

**Comment by fredden** [comment]
Questions for @jrfnl:
- Do you want me to rename `src/Standards/Squiz/Tests/WhiteSpace/OperatorSpacingUnitTest.inc` to `src/Standards/Squiz/Tests/WhiteSpace/OperatorSpacingUnitTest.1.inc` (and `...inc.fixed`) in this pull request?
- Would you like me to rebase this to reduce the number of back/forth commits, or would you prefer to squash-merge this pull request when ready?

**Comment by jrfnl** [comment]
@fredden 

> `switch` statements do not require a `default` case. There is a `return` statement immediately following the `switch`.

🤦🏻‍♀️ I'm sorry, I completely missed those `return` statements after the `switch` when reviewing. My bad.

> These files were very useful during debugging / development. I have removed them as per your preference.

Not so much my preference, as it is a convention in this repo.

Having said that, I didn't realize until your remark that if a `.fixed` file is missing and fixes are being made for a test case file, that the test suite would not flag this.

I've had a look at this and prepared a fix, but I think this will need to wait until PHPCS 4.0 as there are quite a few external standards which use the PHPCS native test framework and changing the framework to flag a missing `fixed` file as a test failure will likely break builds for a number of those.

Aside from that, if I run the PHPCS native test suite with that extra failure, it flags up 5 missing `.fixed` files in PHPCS itself and those will need to be sorted too before the extra failure check could be turned on.

I've opened issue #299 and #300 to address these.


> I was focusing on the PSR12 standard. I have now moved the fix to the Squiz standard and added a test there to catch the same problem.
>
> I didn't want to modify anything outside of the scope I was looking at: PSR12. Given the above, the check is now in the isOperator() check.

Thanks for moving the check. Looking at it as it currently is - I still have a question:
* Why did you put the check at the _end_ of the `isOperator()` method ?
    I would expect it at the very beginning before any of the other checks as determining whether the operator is an operator which should be handled, is irrelevant if the sniff detects a live coding situation.

> I don't have a good justification for this. I've applied the suggested change.

👍🏻 

Might be good to safeguard this by adding one or two extra new lines at the end of the file in the "live coding" test case files ?

> This sounds like a good improvement that could be made to the sniff. I don't think it's in scope here, but I've make a code change to cover this to address this feedback.

Not all questions need to be addressed, sometimes they are just questions. 😉 

I've looked at the fix you applied now and am going to ask you to remove that commit. As I said before, there is a _lot_ wrong with that sniff and the extra code change you've applied now do not mitigate this. Not your fault, it's just a really bad sniff.

Case in point: the `$stackPtr + 1` may not exist, the `findNext()` on line 49 could return `false` which is not being checked etc

I will address this (and a dozen or more other bugs) in that complete rewrite PR.


> Do you want me to rename `src/Standards/Squiz/Tests/WhiteSpace/OperatorSpacingUnitTest.inc` to `src/Standards/Squiz/Tests/WhiteSpace/OperatorSpacingUnitTest.1.inc` (and `...inc.fixed`) in this pull request?
>  - See 25883e8b9d9745751eb7edb0d13a016775f98689. I'm happy to open a separate pull request for this, or cherry-pick that commit into this branch / pull request.

Yes please, that should be in a commit prior to the fix commit.

> Would you like me to rebase this to reduce the number of back/forth commits, or would you prefer to squash-merge this pull request when ready?

I would not squash merge this PR as it contains file renames and those don't always get tracked correctly when squashed in with other changes _(and it is much more helpful when looking at the file history if they do get tracked correctly)_.

I would prefer the actual "fix" commits to be squashed to one commit per fix.

I'm happy to do this myself when the PR is ready for commit though, so you don't need to worry about this if you don't want to.

If you do end up rebasing and reorganizing the commits yourself, I have one extra suggestion: prefix commits with the (short) name of the sniff they relate to, like "PSR12/DeclareStatement: rename test file", that make the commit history much more easy to scan.

Does that answer your questions ?

**Comment by fredden** [comment]
> * Why did you put the check at the _end_ of the `isOperator()` method ?
>   I would expect it at the very beginning before any of the other checks as determining whether the operator is an operator which should be handled, is irrelevant if the sniff detects a live coding situation.

The other checks in the function are less expensive. Calling a function is more expensive than using a built-in like `isset()` or hash look-ups and equality operators (`===`). I can move this to the start of the function if you would prefer. (Putting it at the start makes sense from a logical point of view, so I don't have objections if you would like this at the start instead.)

> Might be good to safeguard this by adding one or two extra new lines at the end of the file in the "live coding" test case files ?

There are a lot of variations which could be covered by tests. These are likely to identify more problems with the sniff(s) though. For example, `<?php declare(;` produces a number of errors which don't make sense (like "Expected no space between opening parenthesis and directive in a declare statement"). I'd prefer to have these in another pull request - perhaps the "rewrite sniff" work that you've alluded to.

> > Do you want me to rename [...]
> Yes please, that should be in a commit prior to the fix commit.

I've added this commit to the pull request. I've put it at the end so that reviewing is kept easy. (GitHub has a nice feature to show what has changed since "my last review," which breaks when history is rewritten.) We can change the order of commits when we're ready to re-base.

**Comment by jrfnl** [comment]
@fredden 

> > * Why did you put the check at the _end_ of the `isOperator()` method ?
> >   I would expect it at the very beginning before any of the other checks as determining whether the operator is an operator which should be handled, is irrelevant if the sniff detects a live coding situation.
> 
> The other checks in the function are less expensive. Calling a function is more expensive than using a built-in like `isset()` or hash look-ups and equality operators (`===`).

Agreed, but the other checks don't just do hash look-ups, they also include `array_*()` function calls (more expensive) and `a call to `File::isReference()` (also more expensive as it is a) a function call too and b) contains more logic).

> I can move this to the start of the function if you would prefer. (Putting it at the start makes sense from a logical point of view, so I don't have objections if you would like this at the start instead.)

Yes please, I think it makes sense to move it to the start.

> > Might be good to safeguard this by adding one or two extra new lines at the end of the file in the "live coding" test case files ?
> 
> There are a lot of variations which could be covered by tests. These are likely to identify more problems with the sniff(s) though. For example, `<?php declare(;` produces a number of errors which don't make sense (like "Expected no space between opening parenthesis and directive in a declare statement"). I'd prefer to have these in another pull request - perhaps the "rewrite sniff" work that you've alluded to.

That's totally fair. I will keep it in mind for the refactor PR. (and yes, I need to find time to finish that at some point, the reason I didn't finish it before is largely cause so much needs to change that it becomes very difficult to do this in clean atomic commits and to decide what commits are needed and in which order)


> > > Do you want me to rename [...]
> > > Yes please, that should be in a commit prior to the fix commit.
> 
> I've added this commit to the pull request. I've put it at the end so that reviewing is kept easy. (GitHub has a nice feature to show what has changed since "my last review," which breaks when history is rewritten.) We can change the order of commits when we're ready to re-base.

👍🏻

**Comment by fredden** [comment]
Thanks for all the feedback. I think we're at a position where the final delta in this pull request is acceptable. If that's the case, shall I now rewrite the history so that the commits are cleaner?

File: src/Standards/Squiz/Sniffs/WhiteSpace/OperatorSpacingSniff.php
Comment (jrfnl) [inline src/Standards/Squiz/Sniffs/WhiteSpace/OperatorSpacingSniff.php:328]: Why are you using `Tokens::$emptyTokens` instead of `T_WHITESPACE` here ?

Comment (fredden) [reply src/Standards/Squiz/Sniffs/WhiteSpace/OperatorSpacingSniff.php:328]: Because `<?php $a = /* one */ 1;` is a valid assignment, and `<?php $a = /* one */` is a parse error.

This highlights another test case which should probably be added: `<?php $a = ;` Would you like me to add this test in scope of this pull request, or is that something for another time? I've checked just now and this branch / pull request already handles these scenarios.

Comment (jrfnl) [reply src/Standards/Squiz/Sniffs/WhiteSpace/OperatorSpacingSniff.php:328]: > Because `<?php $a = /* one */ 1;` is a valid assignment, and `<?php $a = /* one */` is a parse error.

True, but that's irrelevant for these sniffs. In both cases, the sniffs check that there is one space between the operator and the comment and it can do that check, even if there is nothing after the comment.

Then again... if there is a parse error, the "skip default values in function declarations + skip declare statements" check at the start of the `isOperator()` method is unreliable, which is really the heart of the problem which this PR is trying to solve.

This also basically confirms that the parse error check needed to be moved up.

Looking at the sniffs again - the sniffs already contain defensive coding against parse errors and realistically that defensive coding is sufficient, except for the "skip default values in function declarations + skip declare statements" check - which is limited to only apply to two operator tokens, while the sniff checks for a lot more tokens.

So maybe the new defensive coding being added should only be applied to the "skip default values in function declarations + skip declare statements" check ? in which case, using `Tokens::$emptyTokens` instead of `T_WHITESPACE` would be correct.

@fredden What do you think ?


> This highlights another test case which should probably be added: `<?php $a = ;` Would you like me to add this test in scope of this pull request, or is that something for another time? I've checked just now and this branch / pull request already handles these scenarios.

I think we should leave that for another PR and probably check for fixer conflicts on that code snippet with sniffs which check the spacing before a semi-colon, like the `Squiz.WhiteSpace.SemicolonSpacing` sniff.

Comment (fredden) [reply src/Standards/Squiz/Sniffs/WhiteSpace/OperatorSpacingSniff.php:328]: I've looked through <https://www.php.net/manual/en/language.operators.php>. Nearly all operators require something to their right in order to be valid PHP. The exceptions here are the post-increment and post-decrement operators (`$a++` and `$a--`), and the second tick operator (which closes the first). However, in those cases, we still need a non-comment token (eg `;` or `?>`) in order to not be a parse error.

With this in mind, I think it's fair to say that a particular token (which might be an operator) is part of a parse error if it's not followed by a non-comment token. I'd like to keep the parse-error detection as-is, and at the top of the method.

I tried moving the check inside the if-block that you mentioned, and the current tests in this pull request still pass. However the parse error `<?php $a+/* comment */` is no longer detected as a parse error and the sniff raises errors. Shall I add this as a test case here too?

Comment (jrfnl) [reply src/Standards/Squiz/Sniffs/WhiteSpace/OperatorSpacingSniff.php:328]: @fredden The thing is - the sniff is basically split in two parts:
1. Check spacing _before_ the operator.
2. Check spacing _after_ the operator.

The "is operator?" check is intended to check if the sniff needs to run at all.

The difference between what you are proposing and what I'm suggesting is this:
* With the parse error check based on "code after" at the start of the "is operator?" check, the sniff bows out early for a parse error and won't run at all during live coding.
* With the parse error check limited to the "declare statement/function declaration" check within the "is operator?" check, the sniff would still run the "spacing _before_ the operator" part during live coding.
    The sniff `process()` method already contains defensive coding to bow out between the _before_ and _after_ check if there is nothing after.

> However the parse error <?php $a+/* comment */ is no longer detected as a parse error and the sniff raises errors. 

And so it should.
The spacing _before_ the `+` can be checked and the spacing _after_ can be checked as well (as there is a comment).

Try a code sample like this: `<?php $a+`.
Here you will see that the sniff checks only the spacing before, but not the spacing after (with the original code) and that is fine as the spacing before the operator can be checked already, while the spacing after cannot be checked due to the parse error.
With the change as you are proposing now, that snippet would be ignored completely in live coding until the statement would be complete.

Does that make it clearer why I'm suggesting to limit the additional defensive coding to the "declare statement/function declaration" check within the "is operator?" check ?

Comment (jrfnl) [reply src/Standards/Squiz/Sniffs/WhiteSpace/OperatorSpacingSniff.php:328]: We discussed this further off-GitHub and decided on a way forward to get this in a merge-ready state.

File: src/Standards/Squiz/Tests/WhiteSpace/OperatorSpacingUnitTest.1.inc.fixed
Comment (fredden) [inline src/Standards/Squiz/Tests/WhiteSpace/OperatorSpacingUnitTest.1.inc.fixed:185]: I'm undecided about this change. The `?` and `:` symbols _are_ operators in this context, but the sniff is specifically skipping over the whole function parameter list. From what I can tell from the comments, the sniff has always intended to skip these, but the implementation didn't work that way until now.

Comment (jrfnl) [reply src/Standards/Squiz/Tests/WhiteSpace/OperatorSpacingUnitTest.1.inc.fixed:185]: Good point. Hm... in that case, maybe we should just skip over declare statements and keep the pre-existing logic for function declarations only ?

I also wonder if we could only apply that logic for the `T_EQUAL` for default values ? I have a feeling the `T_MINUS` in the condition is an outdated condition, which has been superseded by improved handling of unary tokens in the later part of the `isOperator()` method.

**Review by jrfnl — APPROVED** [review]
Thanks for making those updates and reorganizing this branch @fredden, all looks good now.

I'll fix up that alignment issue and will merge it now.

File: src/Standards/Squiz/Sniffs/WhiteSpace/OperatorSpacingSniff.php
Comment (fredden) [reply src/Standards/Squiz/Sniffs/WhiteSpace/OperatorSpacingSniff.php]: When I pasted this in, my editor prefixed the lines with `* `, so it ended up with `* *     [...] called again on the [...]`. I removed the `*`, but it seems I missed the extra space. This was not identified when I ran `phpcs` over the codebase (which I do before committing). Perhaps there's a bug / oversight in the rule-set?

**Comment by jrfnl** [comment]
@michalbundyra A search showed me the WebImpress standard extends the Squiz/OperatorSpacing sniff. You may want to have a look at this PR and update the sniff if needed.

**Comment by michalbundyra** [comment]
Thanks, @jrfnl, will check it out

File: src/Standards/Squiz/Sniffs/WhiteSpace/OperatorSpacingSniff.php
Comment (jrfnl) [reply src/Standards/Squiz/Sniffs/WhiteSpace/OperatorSpacingSniff.php]: I seem to remember working on a fix for this years ago and that there was some discussion on what the fix should look like... Pff.. I probably still have that branch somewhere, though it may have gotten lost on an old laptop (I have a feeling I worked on this while travelling).

Let's park it for when the docs sniff get revisited in their entirety.

**Pull request merged**
